#Trabalho prático # 6
##Introdução à Computação Numérica

###Tópico: Cálculo de zeros de funções

---

Nome: Riquelme Freitas Gomes

DRE: 120032785


---

**Instruções:** 

**Preencha o notebook** com seus dados, seus códigos e **rode todos os códigos** 

**Inclua  um relatório, na parte final do notebook,** com suas conclusões, respostas, observações e comentários relacionados com os tópicos e questões abordadas nas tarefas.  

**A elaboração dos códigos pode ser realizada em duplas, mas o resto do trabalho é individual.**

*Se você fez o seu trabalho em dupla inclua essa informação no relatório. Além disso, se você  também recebeu ajuda de um(a) colega/amigo(a) ou consultou/usou informações da internet ou outras fontes, coloque referências à elas no seu relatório.*


**Entrega do trabalho:**

*Data de entrega: 09/janeiro/2023* 

---


## Tarefa #1:

Desenvolver códigos, na forma de funções, que implementem os seguintes métodos:

a) Método da bissecção. 

b) Método de Newton-Raphson.  

c) Método da secante.

Sugestão: implemente os métodos na forma de funções que devolvem na saída uma mensagem de erro ou o valor aproximado do zero procurado. Evite a possibilidade de laços infinitos considerando um número máximo de iterações como parâmetro de entrada.

In [161]:
# a) Função que implementa o método da bissecção

def bisseccao(f, a, b, tol, Niter):

  k = 1
  fa = f(a)

  while(k <= Niter):
    p = a + ((b - a) / 2)
    fp = f(p)
    if(fp == 0 or ((b - a) / 2) < tol):
      return p, k
    if(fa * fp > 0):
      a = p
      fa = fp
    else:
      b = p

    k = k + 1
  
  print("Tolerância não atingida")

In [162]:
# b) Função que implementa o método de Newton-Raphson

def newton(f, derivada, p0, tol, Niter):
  
  k = 1
  while(k <= Niter):
    p = p0 - (f(p0) / derivada(p0))

    if(abs(p - p0) < tol):
      return p, k
    
    p0 = p
    k = k + 1
  
  print("Tolerância não atingida")

In [163]:
# c) Função que implementa o método da secante

def secante(f, p0, p1, tol, Niter):
  
  k = 1
  while(k <= Niter):
    p = p1 - ((f(p1)*(p1 - p0)) / (f(p1) - f(p0)))

    if(abs(p - p1) < tol):
      return p, k
    
    p0 = p1
    p1 = p
    k = k + 1
  
  print("Tolerância não atingida")

## Tarefa #2: 

a) A equação de van der Waals 
           		$$\left(p+\frac{a}{v^2} \right)(v-b)=R T,$$

onde $p$ é a pressão, $v$ representa o volume molar, $a$ e $b$ são constantes empíricas que dependem do gás,  $R$ representa a constante universal dos gases e $T$ é a temperatura absoluta. 
Determine os volumes molares do dióxido de carbono a uma temperatura  absoluta $T=300\,K$ e pressão de $1$, $10$, $50$ e $100 \,atm$, respectivamente. 
Os valores das constantes são $R=0.082054 \,L\, atm/(mol K)$, $a=3.592 \,L^2 atm/mol^2$ e $b=0.04267 L/mol$.

Use dois dos métodos implementados considerando uma tolerância $\varepsilon=5\cdot10^{-6}$.

b) Compare os resultados dos dois métodos usados, considerando os valores obtidos e a quantidade de iterações realizadas. Escreva suas observações no relatório.


In [169]:
# a)

pressoes = [1, 10, 50, 100]  # valor da pressão
a, b = 3.592, 0.04267  # constantes empíricas
R = 0.082054  # constante universal dos gases
T = 300  # temperatura absoluta

def van_der_waals(v):
  return (p + (a / v**2)) * (v - b) - (R * T)

tol = 5 * (10**(-6))
intervalo = [0.01, 30]
Niter = 25

for i in range(len(pressoes)):
  p = pressoes[i]
  vb, kb = bisseccao(van_der_waals, intervalo[0], intervalo[1], tol, Niter)
  p0 = vb
  p1 = vb + 1
  vs, ks = secante(van_der_waals, p0, p1, tol, Niter)

  print(f"PRESSÃO DE {p} ATM:")
  print(f"[Bissecção] Volume molar: {vb}\n[Secante] Volume molar: {vs}\n")

PRESSÃO DE 1 ATM:
[Bissecção] Volume molar: 24.512588862180714
[Secante] Volume molar: 24.51258812844154

PRESSÃO DE 10 ATM:
[Bissecção] Volume molar: 2.3544953119754783
[Secante] Volume molar: 2.3544955806028005

PRESSÃO DE 50 ATM:
[Bissecção] Volume molar: 0.3584243643283844
[Secante] Volume molar: 0.35842148005033103

PRESSÃO DE 100 ATM:
[Bissecção] Volume molar: 0.07951040863990785
[Secante] Volume molar: 0.07951060932893776



In [170]:
# b)

for i in range(len(pressoes)):
  p = pressoes[i]
  vb, kb = bisseccao(van_der_waals, intervalo[0], intervalo[1], tol, Niter)
  p0 = vb
  p1 = vb + 1
  vs, ks = secante(van_der_waals, p0, p1, tol, Niter)

  print(f"PRESSÃO DE {p} ATM:")
  print(f"[Bissecção] Volume molar: {vb}\tNumero de iterações: {kb}")
  print(f"[Secante] Volume molar: {vs}\tNumero de iterações: {ks}")
  print(f"Módulo da diferença: {abs(vb - vs)}\n")

PRESSÃO DE 1 ATM:
[Bissecção] Volume molar: 24.512588862180714	Numero de iterações: 23
[Secante] Volume molar: 24.51258812844154	Numero de iterações: 2
Módulo da diferença: 7.33739174307857e-07

PRESSÃO DE 10 ATM:
[Bissecção] Volume molar: 2.3544953119754783	Numero de iterações: 23
[Secante] Volume molar: 2.3544955806028005	Numero de iterações: 2
Módulo da diferença: 2.6862732216414997e-07

PRESSÃO DE 50 ATM:
[Bissecção] Volume molar: 0.3584243643283844	Numero de iterações: 23
[Secante] Volume molar: 0.35842148005033103	Numero de iterações: 2
Módulo da diferença: 2.884278053394418e-06

PRESSÃO DE 100 ATM:
[Bissecção] Volume molar: 0.07951040863990785	Numero de iterações: 23
[Secante] Volume molar: 0.07951060932893776	Numero de iterações: 2
Módulo da diferença: 2.006890299083297e-07



##Tarefa #3:  

Observe que o número $\pi$ é um zero das funções:

i) $f(x)= \cos(x)+1$

ii) $f(x)= \mathrm{sen}(x)$

a) Use sua implementação do método de Newton-Raphson para obter aproximações de considerando as funções acima e uma tolerância $\varepsilon =10^{-16}$.

b) Compare os resultados dos itens i) e ii), considerando os erros das aproximações obtidas com relação ao valor "exato" dado pelo Python no módulo math e a quantidade de iterações efetuadas. Escreva suas observações no relatório.

c) Explique o que foi observado no item b) no seu relatório.

In [164]:
# a)
import math 

# i)
def f(x):
  return math.cos(x) + 1

def derivada_f(x):
  return -1*math.sin(x)

# ii)
def f2(x):
  return math.sin(x)

def derivada_f2(x):
  return math.cos(x) 

p0 = 3.14
tol = math.pow(10, -16)
Niter = 20
aproximacoes = []
qtd_iteracoes = []

p, k = newton(f, derivada_f, p0, tol, Niter)
aproximacoes.append(p)
qtd_iteracoes.append(k)
print(f"Função i)\nAproximação de pi: {p}\n")

p, k = newton(f2, derivada_f2, p0, tol, Niter)
aproximacoes.append(p)
qtd_iteracoes.append(k)
print(f"Função ii)\nAproximação de pi: {p}")

Função i)
Aproximação de pi: 3.141592650867387

Função ii)
Aproximação de pi: 3.141592653589793


In [165]:
# b) 
import math

valor_exato = math.pi
for i in range(len(aproximacoes)):
  valor_aproximado = aproximacoes[i]
  
  print(f"Aproximação de pi: {valor_aproximado}\nValor exato de pi: {valor_exato}")
  print(f"Erro absoluto: {abs(valor_exato - valor_aproximado)}")
  print(f"Numero de iterações: {qtd_iteracoes[i]}\n")

Aproximação de pi: 3.141592650867387
Valor exato de pi: 3.141592653589793
Erro absoluto: 2.7224063003927768e-09
Numero de iterações: 19

Aproximação de pi: 3.141592653589793
Valor exato de pi: 3.141592653589793
Erro absoluto: 0.0
Numero de iterações: 3



# **Relatório**

Nome: Riquelme Freitas Gomes \
DRE: 120032785

\
## **Trabalho prático #6**
Os códigos presentes neste documento são todos de minha autoria e para implementar alguns deles consultei o seguinte material: https://www.dropbox.com/s/zypuxsl68622qf9/slides-Zeros%20de%20fun%C3%A7%C3%B5es.pdf?dl=0

\
## **Tarefa #1:**
Nessa tarefa, desenvolvi códigos, na forma de funções, que implementem os seguintes métodos:

### **a) Método da bissecção.**

### **b) Método de Newton-Raphson.**

### **c) Método da secante.**

\
## **Tarefa #2:**

### **a)**
Nesse item, determinei os volumes molares do dióxido de carbono dadas as condições e valores presentes na questão e considerando valores de pressão de  1,  10,  50  e  100 atm. Para isso, utilizei a equação de van der Waals e usei dois dos meus métodos implementados, foram eles o método da bissecção e o método da secante. Em ambos considerei uma tolerância $\varepsilon=5\cdot10^{-6}$ e um $N_{iter}$ (número máximo de iterações) igual a 25. Para o método da bissecção convencionei o intervalo $[0.01, 30]$ e para o método da secante utilizei a aproximação obtida com o método anterior para determinar os valores de $p0$ e $p1$ (aproximações iniciais). Obtive os seguintes resultados: 

```
PRESSÃO DE 1 ATM:
[Bissecção] Volume molar: 24.512588862180714
[Secante] Volume molar: 24.51258812844154

PRESSÃO DE 10 ATM:
[Bissecção] Volume molar: 2.3544953119754783
[Secante] Volume molar: 2.3544955806028005

PRESSÃO DE 50 ATM:
[Bissecção] Volume molar: 0.3584243643283844
[Secante] Volume molar: 0.35842148005033103

PRESSÃO DE 100 ATM:
[Bissecção] Volume molar: 0.07951040863990785
[Secante] Volume molar: 0.07951060932893776
```

\
### **b)**
Nesse item, comparei os resultados dos dois métodos usados, considerando os valores obtidos e a quantidade de iterações realizadas. Estes foram os resultados: 

```
PRESSÃO DE 1 ATM:
[Bissecção] Volume molar: 24.512588862180714	Numero de iterações: 23
[Secante] Volume molar: 24.51258812844154	   Numero de iterações: 2
Módulo da diferença: 7.33739174307857e-07

PRESSÃO DE 10 ATM:
[Bissecção] Volume molar: 2.3544953119754783	Numero de iterações: 23
[Secante] Volume molar: 2.3544955806028005  	Numero de iterações: 2
Módulo da diferença: 2.6862732216414997e-07

PRESSÃO DE 50 ATM:
[Bissecção] Volume molar: 0.3584243643283844	Numero de iterações: 23
[Secante] Volume molar: 0.35842148005033103	 Numero de iterações: 2
Módulo da diferença: 2.884278053394418e-06

PRESSÃO DE 100 ATM:
[Bissecção] Volume molar: 0.07951040863990785	Numero de iterações: 23
[Secante] Volume molar: 0.07951060932893776  	Numero de iterações: 2
Módulo da diferença: 2.006890299083297e-07
```

Observei que, para qualquer valor da pressão, o método da bissecção sempre encontrou a aproximação do volume molar com 23 iterações, enquanto o método da secante o fez com apenas 2 iterações. Em todos os casos os volumes molares encontrados por cada método apresentaram valores bem próximos e, consequentemente, uma diferença pequena. Nesse contexto, apesar do método da bissecção funcionar sob condições bastante gerais e o método da secante necessitar de boas aproximações iniciais, o método da secante se mostrou mais eficiente, uma vez que convergiu a um valor necessitando de um número muito menor de iterações.

\
## **Tarefa #3:**

### **a)**
Nesse item, usei minha implementação do método de Newton-Raphson para obter aproximações de $\pi$ considerando as funções $i)$ e $ii)$ presentes na questão e suas respectivas derivadas. Considerei uma tolerância $\varepsilon =10^{-16}$, um $p_0$ (aproximação inicial) de $3.14$ e um $N_{iter}$ (número máximo de iterações) igual a 20. Obtive os seguintes resultados: 

```
Função i)
Aproximação de pi: 3.141592650867387

Função ii)
Aproximação de pi: 3.141592653589793
```

\
### **b)**
Nesse item, comparei os resultados atingidos com as funções $i)$ e $ii)$ no item anterior, considerando os erros das aproximações obtidas com relação ao valor "exato" dado pelo Python no módulo math e a quantidade de iterações efetuadas. Esses foram os resultados:

```
Função i)
Aproximação de pi: 3.141592650867387
Valor exato de pi: 3.141592653589793
Erro absoluto: 2.7224063003927768e-09
Numero de iterações: 19

Função ii)
Aproximação de pi: 3.141592653589793
Valor exato de pi: 3.141592653589793
Erro absoluto: 0.0
Numero de iterações: 3
```

### **c)**
Utilizando a função $i)$ encontrei o valor aproximado para $\pi$ como sendo de $3.141592650867387$ e com um erro absoluto pequeno e igual a $2.7224063003927768e^{-09}$, realizando 19 iterações. Já com a função $ii)$, com apenas 3 iterações, o valor de $\pi$ encontrado foi de $3.141592653589793$, que é o mesmo valor "exato" dado pelo Python no módulo math. Por consequência, tal aproximação apresentou um erro absoluto igual a 0. A partir dessa comparação, é possível constatar que, nessas condições, observamos melhores resultados quando utilizamos a implementação do método de Newton-Raphson para obter aproximações de $\pi$ considerando a função $ii)$, pois além de executar a aproximação em um número muito menor de iterações, atingiu o valor "exato" que convencionamos para $\pi$.